### PR TITLE
fix(client): fluent api nulls

### LINKED ISSUE #5715: Wrong types for fluent API when used along with `rejectOnNotFound`
Hey, 

I just want to report issue I just found out.

Schema:
```prisma
datasource db {
  provider = "sqlite"
  url      = "file:./build.db"
}

generator client {
  provider = "prisma-client-js"
  output   = "./dist"
}

model Composer {
  id        String        @id @default(cuid())
  pieces    Piece[]
  createdAt DateTime      @default(now())
  updatedAt DateTime      @updatedAt
}

model Piece {
  id         String        @id @default(cuid()
  composer   Composer      @relation(fields: [composerId], references: [id])
  composerId String
  createdAt  DateTime      @default(now())
  updatedAt  DateTime      @updatedAt
}
```

See code below with comments:

```tsx
// OK -> piece: Piece
const piece = await prisma.piece.findUnique({
  where: {
    id: parent.id,
  },
  rejectOnNotFound: true,
});

// OK -> piece: Piece | null
const piece = await prisma.piece.findUnique({
  where: {
    id: parent.id,
  },
});

// OK -> composer: Composer | null
const composer = await prisma.piece
  .findUnique({
    where: {
      id: parent.id,
    },
  })
  .composer();

// BAD -> composer: Composer | null, should be composer: Composer
const composer = await prisma.piece
  .findUnique({
    where: {
      id: parent.id,
    },
    rejectOnNotFound: true,
  })
  .composer();

// To fix this we need to do this:
// OK -> composer: Composer
const composer = await prisma.piece
  .findUnique({
    where: {
      id: parent.id,
    },
    select: { composer: true },
    rejectOnNotFound: true,
  }).then(piece=>piece.composer)
```

### LINKED ISSUE #9138: A model with relation to another model via non-null foreign key has nullable type
### Bug description

See  #9137

Take the following schema adapted from the starter project:
```prisma
model Post {
  id        Int     @id @default(autoincrement())
  content   String
  author    User   @relation(fields: [authorId], references: [id])
  authorId  Int
}

model User {
  id    Int     @id @default(autoincrement())
  name  String
  posts Post[]
}
```

This results in a TS client that has following return types:
```typescript
let author = await prisma.post.findUnique({
  rejectOnNotFound: true,
  where: {id: 1}
})        // type: Post
.author() // type: User | null
```

We can see that the return value of `findUnique` call will have non-null type `Post` since `rejectOnNotFound` was set to `true`.
But the subsequent call to `authors()` yields `User | null` type. This to me seems like an erroneous union with null since we've specified non-null foreign key to the `User` model.

### How to reproduce

Define the schema that is shown in the example and generate a js/ts client.

### Expected behavior

The return type of the `author()` call should be `User`, rather than `User | null`.

### Prisma information

Refer to the bug description.

### Environment & setup

- OS: Arch
- Database: PostgreSQL
- Node.js version: v16.8.0

(I believe this issue is independent of the environment)

### Prisma Version

```
prisma                : 2.30.2
@prisma/client        : 2.30.2
Current platform      : debian-openssl-1.1.x
Query Engine (Binary) : query-engine b8c35d44de987a9691890b3ddf3e2e7effb9bf20 (at node_modules/@prisma/engines/query-engine-debian-openssl-1.1.x)
Migration Engine      : migration-engine-cli b8c35d44de987a9691890b3ddf3e2e7effb9bf20 (at node_modules/@prisma/engines/migration-engine-debian-openssl-1.1.x)
Introspection Engine  : introspection-core b8c35d44de987a9691890b3ddf3e2e7effb9bf20 (at node_modules/@prisma/engines/introspection-engine-debian-openssl-1.1.x)
Format Binary         : prisma-fmt b8c35d44de987a9691890b3ddf3e2e7effb9bf20 (at node_modules/@prisma/engines/prisma-fmt-debian-openssl-1.1.x)
Default Engines Hash  : b8c35d44de987a9691890b3ddf3e2e7effb9bf20
Studio                : 0.422.0
```

### LINKED ISSUE #10656: Wrong type when retrieving multiple related records using fluent API
### Bug description

Similar to #5715, but slightly different.

### How to reproduce

```ts
// BAD -> pieces: Piece[], should be pieces: Piece[] | null
const pieces = await prisma.composer
  .findUnique({
    where: {
      id: composer.id,
    },
  })
  .pieces();

// If the composer is not found, it will be null, even though its type is Piece[].
// TypeError: Cannot read property 'length' of null
consoloe.log(pieces.length);
```

### Expected behavior

`pieces` must be of type `Piece[] | null` , or `.pieces()` must be available only if `rejectOnNotFound: true`.

### Prisma information

```prisma
model Composer {
  id        String        @id @default(cuid())
  pieces    Piece[]
  createdAt DateTime      @default(now())
  updatedAt DateTime      @updatedAt
}

model Piece {
  id         String        @id @default(cuid())
  composer   Composer      @relation(fields: [composerId], references: [id])
  composerId String
  createdAt  DateTime      @default(now())
  updatedAt  DateTime      @updatedAt
}
```

### Environment & setup

- OS: Mac OS
- Database: PostgreSQL
- Node.js version: 16.5.0

### Prisma Version

```
3.6.0
```

### LINKED ISSUE #10687: Fluent API produces runtime error
### Bug description

In the following example, when `email` does not exist, the result of `links()` is `null` but the types suggest that it's `Link[]`.

### How to reproduce

```ts
import { PrismaClient } from ".prisma/client";

(async () => {
  const prisma = new PrismaClient();

  await prisma.user.create({
    data: {
      email: Date.now() + "@gmail.com",
    },
  });

  const links = await prisma.user
    .findUnique({
      where: { email: "invalid" },
    })
    .links();

  // type: Link[], runtime: null
  console.log("links", links);

  prisma.$disconnect();
})();

```

### Expected behavior

Type should be `Link[] | null`

### Prisma information

```prisma
generator client {
  provider = "prisma-client-js"
  output =  "../node_modules/.prisma/client"
}

datasource db {
  provider = "sqlite"
  url      = "file:./dev.db"
}

model Link {
  id        String   @id @default(uuid())
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt
  url       String
  shortUrl  String
  User      User?    @relation(fields: [userId], references: [id])
  userId    String?
}

model User {
  id        String   @id @default(uuid())
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt
  name      String?
  email     String   @unique
  links     Link[]
}

```

### Environment & setup

- OS: Ubuntu
- Database: SQLite
- Node.js version: 14.16

### Prisma Version

```
3.6.0
```

### PR DESCRIPTION
**Pull request by millsp**
Fixes https://github.com/prisma/prisma/issues/10687
Fixes https://github.com/prisma/prisma/issues/9138
Fixes https://github.com/prisma/prisma/issues/10656

### CODE REVIEW

**Review by danstarns — COMMENTED** [review]
Is there a way we could add a regression test for this? Maybe something in the functional suite?

**Comment by millsp** [comment]
> Is there a way we could add a regression test for this? Maybe something in the functional suite?

Thanks for the suggestion, I have added them and fixed some cases I've missed along the way.

**Comment by millsp** [comment]
I'm now doing type assertions in the functional test setup thanks to your suggestion Dan.

File: packages/client/tests/functional/fluent-api-null/tests.ts [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+declare let prisma: PrismaClient
+
+const nonExistingId = randomBytes(12).toString('hex')
+
+testMatrix.setupTestSuite((suiteConfig, suiteMeta) => {
+  test('findFirst', async () => {
+    const result = await prisma.resource.findFirst().children()
+
+    expect(result).toBeNull()
+    ;() => Test.checks([Test.check<U.Has<typeof result, null>, 1, Test.Pass>()])
```
[the remark this answers is not in the document]
Comment (millsp) [reply packages/client/tests/functional/fluent-api-null/tests.ts:17]: Nice, I'll add that

**Review by SevInf — APPROVED** [review]

**Pull request merged**